### PR TITLE
【Hackathon 4 No52】add fp16 for dist

### PR DESCRIPTION
**Pull request by enkilee**
### PR types

Others
### PR changes

Others
### Describe

add fp16 for dist
add docs: https://github.com/PaddlePaddle/docs/pull/5730

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_dist_op.py [hunk 1]
```diff
@@ -113,14 +113,13 @@ def get_reduce_dims(x, y):
         return x_grad, y_grad
 
     def test_check_output(self):
-        self.check_output(check_eager=True)
```
Comment (zhangting2020) [inline python/paddle/fluid/tests/unittests/test_dist_op.py]: 如果不是为了解决某种问题，建议不要修改原始单测的设置。

File: python/paddle/fluid/tests/unittests/test_dist_op.py [hunk 2]
```diff
@@ -190,6 +189,11 @@ def test_api(self):
             np.testing.assert_allclose(dist(x_i, y_i, p), out[0], rtol=1e-05)
 
 
+class TestDistFP16OP(TestDistOp):
+    def init_data_type(self):
+        self.data_type = np.float16
```
Comment (zhangting2020) [inline python/paddle/fluid/tests/unittests/test_dist_op.py]: 这个case只能测到p=0的情况，参考上面的5个case，添加其余的几种场景。

**Comment by enkilee** [comment]
Traceback (most recent call last):
2023-03-22 23:09:51     10	  File "test_dist_op.py", line 119, in test_check_grad
2023-03-22 23:09:51     11	    self.check_grad(
2023-03-22 23:09:51     12	  File "/mnt/paddle/build/python/paddle/fluid/tests/unittests/eager_op_test.py", line 2211, in check_grad
2023-03-22 23:09:51     13	    self.check_grad_with_place(
2023-03-22 23:09:51     14	  File "/mnt/paddle/build/python/paddle/fluid/tests/unittests/eager_op_test.py", line 2387, in check_grad_with_place
2023-03-22 23:09:51     15	    self._assert_is_close(
2023-03-22 23:09:51     16	  File "/mnt/paddle/build/python/paddle/fluid/tests/unittests/eager_op_test.py", line 2184, in _assert_is_close
2023-03-22 23:09:51     17	    self.assertLessEqual(max_diff, max_relative_error, err_msg())
2023-03-22 23:09:51     18	AssertionError: 0.04123 not less than or equal to 0.005 : Operator dist error, Gradient Check On Place(gpu:0) variable X (shape: (4, 1, 4, 8), dtype: float16) max gradient diff 4.122925e-02 over limit 5.000000e-03, the first error element is 36, expected 2.960205e-03, but got 2.838135e-03.

请问，像这种grad check超了的，该如何分析？

**Comment by zhangting2020** [comment]
> Traceback (most recent call last): 2023-03-22 23:09:51 10 File "test_dist_op.py", line 119, in test_check_grad 2023-03-22 23:09:51 11 self.check_grad( 2023-03-22 23:09:51 12 File "/mnt/paddle/build/python/paddle/fluid/tests/unittests/eager_op_test.py", line 2211, in check_grad 2023-03-22 23:09:51 13 self.check_grad_with_place( 2023-03-22 23:09:51 14 File "/mnt/paddle/build/python/paddle/fluid/tests/unittests/eager_op_test.py", line 2387, in check_grad_with_place 2023-03-22 23:09:51 15 self._assert_is_close( 2023-03-22 23:09:51 16 File "/mnt/paddle/build/python/paddle/fluid/tests/unittests/eager_op_test.py", line 2184, in _assert_is_close 2023-03-22 23:09:51 17 self.assertLessEqual(max_diff, max_relative_error, err_msg()) 2023-03-22 23:09:51 18 AssertionError: 0.04123 not less than or equal to 0.005 : Operator dist error, Gradient Check On Place(gpu:0) variable X (shape: (4, 1, 4, 8), dtype: float16) max gradient diff 4.122925e-02 over limit 5.000000e-03, the first error element is 36, expected 2.960205e-03, but got 2.838135e-03.
> 
> 请问，像这种grad check超了的，该如何分析？

从前、反向的实现上去检查，看哪个计算过程对精度损失比较严重

**Comment by luotao1** [comment]
close due to the following PR is merged:
- #50915

**Pull request closed without merging**